### PR TITLE
Support HTTP Proxy for Services.AppAuthentication and ADAL

### PR DESCRIPTION
**Pull request by rlaveycal**
This PR allows HTTP proxies to be properly supported via the IHttpClientFactory interface (https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Providing-an-HttpClient).

The manner for passing this interface into AzureServiceTokenProvider is either by constructor parameter (so that upstream libraries can play nice) or via static property AzureServiceTokenProvider.HttpClientFactory (so that upstream clients can get proxy support without waiting for upstream libraries). This solves the outstanding issues with #4404

i.e. my client application behind a web proxy can now get a secret from KeyVault using a certificate for service principal authentication.

Rebases/replaces #6451

### CODE REVIEW

**Comment by rlaveycal** [comment]
@nonik0 @AlexGhiondea as requested. The unit tests are horrible but I couldn't see any way of checking the internals.

**Comment by nonik0** [comment]
@varunsh-msft could you tag AppAuth and also review/give feedback if you have time?

@rlaveycal Thanks for the tests. I took a look at your new PR and have a few thoughts. Ideally, you could craft your MockHttpClientFactory to create a mock response that ADAL will accept and return a valid response for, then you can validate the mock response/token is as expected. That way you shouldn't have to expose any test hooks to be able to validate the factory was passed down to underlying AdalAuthContext, as well as fully testing the whole scenario by doing the ADAL call with the mock.

If that's not feasible, then another approach to validate could be to set the access level of ASTP's `_selectedAccessTokenProvider` and `_potentialAccessTokenProviders` to internal. Then, after passing the MockHttpClientFactory to the ASTP ctor, the test can access and validate the underlying applicable token providers that they have the mock factory. I believe there should be 4 tests for this, one for passing each connection string to ASTP that will inject the mock factory (see [here](https://docs.microsoft.com/en-us/azure/key-vault/service-to-service-authentication#connection-string-support) for reference), and one for no connection string (so this covers all scenarios where mock is injected).

If you need to, you can take a look at some other tests that are using internal access level for test validation, just look for other internal fields, as other tests are likely referencing those. [This is the line](https://github.com/Azure/azure-sdk-for-net/blob/d03850c2c837ebfa2403ef5d9cd1acda4340abc9/sdk/mgmtcommon/AppAuthentication/Azure.Services.AppAuthentication/Properties/AssemblyInfo.cs#L13) that allow internals to be exposed to our test code (including just in case you were not aware).

**Comment by varunsh-msft** [comment]
Added the label. @nonik0 , please do look into nuget version update. I did not see that change. Thanks!

**Comment by nonik0** [comment]
@azuresdkci @AlexGhiondea ACK after the most recent commit, feel free to merge.

**Comment by tebeco** [comment]
Wouhou, it's finally going to happen :D

**Comment by tebeco** [comment]
I do confirm that for now it is impossible to reach the `Azure KeyVault` :
```cs
                var httpClientHandler = new HttpClientHandler()
                {
                    Proxy = new WebProxy(new Uri(keyVaultOptions.ProxyUri)),
                    SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls,
                    UseProxy = true
                };
                var httpClient = new HttpClient(httpClientHandler);
                var keyVaultClient = new KeyVaultClient(new KeyVaultClient.AuthenticationCallback(authenticationCallback), httpClient);

                var azureKeyVaultConfigurationOptions = new AzureKeyVaultConfigurationOptions(
                    keyVaultOptions.Uri,
                    keyVaultOptions.ClientId,
                    keyVaultOptions.Secret);
                azureKeyVaultConfigurationOptions.Manager = SecretManagerFactory.CreateFromOptions(keyVaultOptions);
                azureKeyVaultConfigurationOptions.Client = keyVaultClient;

                configurationBuilder.AddAzureKeyVault(azureKeyVaultConfigurationOptions);

                async Task<string> authenticationCallback(string authority, string resource, string scope)
                {
                    var authenticationContext = new AuthenticationContext(authority);
                    var clientCredential = new ClientCredential(keyVaultOptions.ClientId, keyVaultOptions.Secret);
                    var result = await authenticationContext.AcquireTokenAsync(resource, clientCredential).ConfigureAwait(false);

                    if (result == null)
                    {
                        throw new InvalidOperationException($"Could not aquired AccessToken for {keyVaultOptions.Uri} with ClientId '{keyVaultOptions.ClientId}'");
                    }

                    return result.AccessToken;
                }
```



As a consumer, a target desgin would be to do :

```cs
                var azureKeyVaultConfigurationOptions = new AzureKeyVaultConfigurationOptions(
                    keyVaultOptions.Uri,
                    keyVaultOptions.ClientId,
                    keyVaultOptions.Secret);
                azureKeyVaultConfigurationOptions.Manager = SecretManagerFactory.CreateFromOptions(keyVaultOptions);
                azureKeyVaultConfigurationOptions.Client = keyVaultClient;
                azureKeyVaultConfigurationOptions.Proxy = new WebProxy(new Uri(keyVaultOptions.ProxyUri)); // <==== this is important to be an IWebProxy to let it open for extensibility

                configurationBuilder.AddAzureKeyVault(azureKeyVaultConfigurationOptions);
```

**Comment by tebeco** [comment]
Also this might seems a stupid question but after reading this :
https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Providing-an-HttpClient

I would love to see a working example of the `AuthenticationContext` setup for a proxy, does the Url  `"https://login.microsoftonline.com/common"` is supposed to be known by consumer ?
That why, as a consumer, i would highly prefer just setting an `IWebProxy` to the existing `AzureKeyVaultConfigurationOptions` that is part of the SDK
and the internal of `configurationBuilder.AddAzureKeyVault` (the same SDK) would setup the factory properly to avoid leak internals to the consumer

As the release of this PR would seem to unlock the underlying blocking issue with proxy, i do understand that a first release of it would unlock us, waiting for an `Option` API

**Comment by tebeco** [comment]
will it be testableright away after @AlexGhiondea merge it ? \o/
I guess i would just have to hit the public static API ?

or does `Microsoft.Azure.KeyVault` will also need to be updated ?

**Comment by nonik0** [comment]
Ping @azuresdkci, @AlexGhiondea -- not sure why this check-enforcer check isn't reporting but this PR should be ready to merge.

@tebeco I can't speak for other SDKs, but AppAuth isn't synchronized with any other releases and they are something we need to initiate manually. Your changes, once merged, would be a part of the next release.

**Comment by tebeco** [comment]
> which will not pass until all other checks that are in progress have completed successfully

If PR are squashed (no idea how this repo merge PR) you could try an empty commit with `git commit --allow-empty -m "Trigger the CI as check-enforcer did not ran"` to see

**Comment by tebeco** [comment]
Is there anyway to contact @AlexGhiondea in other channels as this PR would unlock almost all Azure SDK proxy issue if I understand correctly.

**Comment by tebeco** [comment]
I'm sorry if I sound a bit intrusive here, I don't know what to do to help
This fix will probably help and unlock lots of people

It seems that it only needs approval and merge ?

**Comment by nonik0** [comment]
@AlexGhiondea Could you approve PR please?

**Comment by AlexGhiondea** [comment]
@nonik0 this introduces some new APIs -- I think you should bump the package version.

**Review by AlexGhiondea — APPROVED** [review]

**Comment by tebeco** [comment]
Thx a lot @AlexGhiondea

**Comment by nonik0** [comment]
@tebeco I should be releasing another PR with more AppAuth updates on our end sometime today, and once that's merged I can do a new AppAuth release that includes your changes.

**Comment by tebeco** [comment]
> that includes your changes

What do you have in mind for `your changes`, I'm very very happy to see all these changes being merged to have proxy support ;)
But to be honest, I feel a bit lost ^^ can you be a bit more specific :D ?

**Comment by nonik0** [comment]
Sorry, what I meant by our changes were other unrelated fixes/updates for AppAuth, etc.

You can view the release notes for the [1.5 release](https://www.nuget.org/packages/Microsoft.Azure.Services.AppAuthentication/1.5.0) which was just published and includes your PR.

**Comment by tebeco** [comment]
hello @nonik0 

Thx, I managed to get code connecting properly but the API look like it could be a bit easier.

The sample given [on this pages ](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/wiki/Providing-an-HttpClient) won't work for everyone
Our keyvault are probably not expose and the `authority` is not `https://login.microsoftonline.com/common` but something like `https://login.microsoftonline.com/abcd-102312-acbfd123....` (a `Guid`)

The internal of the SDK know this `authority`, do you think it would be possible just to pass an `IWebProxy` to the SDK ... and that's all
I believe that both team handling Azure KV SDK and ADAL already have all the info and we could just give either an `IWebProxy` and/or `HttpHandler`

Right now the code I have look like:

**Comment by tebeco** [comment]
Just test, got it working 

The Api is very confusing TBH ^^
Lots of code duplication or boilerplate for default behavior

Code look like this so far :

```
        public static void AddCorpoKeyVault(this IConfigurationBuilder configurationBuilder, CorpoKeyVaultOptions corpoKvOptions)
        {
                var httpClientHandler = new HttpClientHandler()
                {
                    Proxy = new WebProxy(corpoKvOptions.ProxyUri),
                    SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls,
                    UseProxy = true
                };
                var httpClient = new HttpClient(httpClientHandler);

                var keyVaultClient = new KeyVaultClient(new KeyVaultClient.AuthenticationCallback(authenticationCallback), httpClient);
                var azureKeyVaultConfigurationOptions = new AzureKeyVaultConfigurationOptions(
                    corpoKvOptions.Uri.AbsoluteUri, // <================================================================ Could this be a Uri ??
                    corpoKvOptions.ClientId,
                    corpoKvOptions.Secret) //Only for testing, else use certificate ^^
                {
                    Manager = foo /* Code here to create a Manager because we use prefix segregation*/,
                    Client = keyVaultClient,
                };

                configurationBuilder.AddAzureKeyVault(azureKeyVaultConfigurationOptions);

/*********************************************************************************************************/
// // This API seems overkill
                async Task<string> authenticationCallback(string authority, string resource, string scope)
                {
                    var myHttpClientFactory = new MyHttpClientFactory(corpoKvOptions); 


                    var authenticationContext = new AuthenticationContext(
                         authority: authority,
                         validateAuthority: true, 
                         tokenCache: new TokenCache(), 
                         httpClientFactory: myHttpClientFactory);

                    var clientCredential = new ClientCredential(corpoKvOptions.ClientId, corpoKvOptions.Secret);
                    var result = await authenticationContext.AcquireTokenAsync(resource, clientCredential).ConfigureAwait(false);

                    if (result == null)
                    {
                        throw new InvalidOperationException($"Could not aquired AccessToken for {corpoKvOptions.Uri} with ClientId '{corpoKvOptions.ClientId}'");
                    }

                    return result.AccessToken;
                }
/*********************************************************************************************************/
            }
        }
```

Then we have to create a class:
The biggest issue here is that it re-use an already existing interface `IHttpClientFactory` that is named just like already existing standard here :
https://github.com/dotnet/extensions/tree/v3.1.4/src/HttpClientFactory/Http/src
this is very very confusing to re-use something existing for the same purpose with different type and signature

```
    public class MyHttpClientFactory : Microsoft.IdentityModel.Clients.ActiveDirectory.IHttpClientFactory
    {
        private readonly HttpClient _httpClient;

        public MyHttpClientFactory(CorpoKeyVaultOptions corpoKvOptions)
        {
            var httpClientHandler = new HttpClientHandler()
            {
                Proxy = new WebProxy(corpoKvOptions.ProxyUri),
                SslProtocols = SslProtocols.Tls12 | SslProtocols.Tls11 | SslProtocols.Tls,
                UseProxy = true
            };
            _httpClient = new HttpClient(httpClientHandler);
        }

        public HttpClient GetHttpClient() => _httpClient;
    }
```

As you may see, aside the proxy and manager, the code is just doing default stuff
One idea would be to only have this:
* No need of the `AuthenticationCallback` if you only use the default values
* No need of the `HttpClient` for default values
* No need of the second `HttpClient` for default values
* No need to create a class to create an `httpclient` with all default values
* no need to handle the `authenticationCallback` by default if you re-use the default
```
        public static void AddCorpoKeyVault(this IConfigurationBuilder configurationBuilder, CorpoKeyVaultOptions corpoKvOptions)
        {
                var proxy = new WebProxy(corpoKvOptions.ProxyUri);

                var keyVaultClient = new KeyVaultClient(proxy);
                var azureKeyVaultConfigurationOptions = new AzureKeyVaultConfigurationOptions(
                    corpoKvOptions.Uri,<================================================ With a Uri instead of string here ^^
                    corpoKvOptions.ClientId,
                    corpoKvOptions.Secret) //Only for testing, else use certificate ^^
                {
                    Manager = foo /* Code here to create a Manager because we use prefix segregation*/,
                    Client = keyVaultClient,
                };
```

Do you think something like this already exists ?
Or is it one of the PR you were talking about ?

**Comment by tebeco** [comment]
should i move my previous comment to a dedicated issue ?

**Pull request merged**
